src/Compilers/CSharp/Test/Semantic/Semantics/NativeIntegerTests.cs

```diff
@@ -6251,5 +6251,191 @@ .maxstack  2
+            constantFolding("nint", "const nint A = -2147483648; const int B = 1;", "A >> B", "-1073741824");
+            constantFolding("nint", "const nint A = -1; const int B = 31;", "A >> B", "-1");
+            constantFolding("nint", "const nint A = -1; const int B = 32;", "A >> B", "-1");
+            constantFolding("nuint", "const nuint A = 0; const int B = 1;", "A >> B", "0");
+            constantFolding("nuint", "const nuint A = 4294967295; const int B = 1;", "A >> B", "2147483647");
+            constantFolding("nuint", "const nuint A = 1; const int B = 31;", "A >> B", "0");
+            constantFolding("nuint", "const nuint A = 1; const int B = 32;", "A >> B", "1");
+
+            constantFolding("nint", "const nint A = -2147483648; const nint B = 0;", "A & B", "0");
+            constantFolding("nint", "const nint A = -2147483648; const nint B = -1;", "A & B", "-2147483648");
+            constantFolding("nint", "const nint A = -2147483648; const nint B = 2147483647;", "A & B", "0");
+            constantFolding("nuint", "const nuint A = 0; const nuint B = 4294967295;", "A & B", "0");
+            constantFolding("nuint", "const nuint A = 2147483647; const nuint B = 4294967295;", "A & B", "2147483647");
+            constantFolding("nuint", "const nuint A = 2147483647; const nuint B = 2147483648;", "A & B", "0");
+
+            constantFolding("nint", "const nint A = -2147483648; const nint B = 0;", "A | B", "-2147483648");
+            constantFolding("nint", "const nint A = -2147483648; const nint B = -1;", "A | B", "-1");
+            constantFolding("nint", "const nint A = 2147483647; const nint B = 2147483647;", "A | B", "2147483647");
+            constantFolding("nuint", "const nuint A = 0; const nuint B = 4294967295;", "A | B", "4294967295");
+            constantFolding("nuint", "const nuint A = 2147483647; const nuint B = 2147483647;", "A | B", "2147483647");
+            constantFolding("nuint", "const nuint A = 2147483647; const nuint B = 2147483648;", "A | B", "4294967295");
+
+            constantFolding("nint", "const nint A = -2147483648; const nint B = 0;", "A ^ B", "-2147483648");
+            constantFolding("nint", "const nint A = -2147483648; const nint B = -1;", "A ^ B", "2147483647");
+            constantFolding("nint", "const nint A = 2147483647; const nint B = 2147483647;", "A ^ B", "0");
+            constantFolding("nuint", "const nuint A = 0; const nuint B = 4294967295;", "A ^ B", "4294967295");
+            constantFolding("nuint", "const nuint A = 2147483647; const nuint B = 2147483647;", "A ^ B", "0");
+            constantFolding("nuint", "const nuint A = 2147483647; const nuint B = 2147483648;", "A ^ B", "4294967295");
+
+            void constantFolding(string opType, string declarations, string expr, string expectedResult, DiagnosticDescription diagnostic = null)
+            {
+                var diagnostics = (diagnostic is null) ?
```

Comment:
Consider switching over a tuple of `(diagnostic, expectedResult)` to make this easier to understand.